File: ios/MyJobPitch/Utils/AppData.swift

//  AppData.swift
//  MyJobPitch
//
//  Created by dev on 12/21/16.
//  Copyright © 2016 myjobpitch. All rights reserved.
//

import UIKit

class AppData: NSObject {
    
    // const
    
    static let apiVersion = 6
    static let production = false
    
    static let greenColor = UIColor(red: 0/255.0, green: 182/255.0, blue: 164/255.0, alpha: 1)
    static let yellowColor = UIColor(red: 1, green: 147/255.0, blue: 0, alpha: 1)
    static let lightGreyColor = UIColor(red: 235/255.0, green: 235/255.0, blue: 235/255.0, alpha: 1)
    static let greyColor = UIColor(red: 214/255.0, green: 214/255.0, blue: 214/255.0, alpha: 1)
    static let darkColor = UIColor(red: 51/255.0, green: 51/255.0, blue: 51/255.0, alpha: 1)
    
    static let DEFAULT_REFRESH_TIME = 30
    static let MESSAGE_REFRESH_TIME = 5
    
    // app data
    
    static var email: String! {
        get {
            return UserDefaults.standard.string(forKey: "email")
        }
        set(newEmail) {
            UserDefaults.standard.set(newEmail, forKey: "email")
            UserDefaults.standard.synchronize()
        }
    }
    
    static var user: User!

    static var hours: [Hours]!
    static var contracts: [Contract]!
    static var sexes: [Sex]!
    static var nationalities: [Nationality]!
    static var sectors: [Sector]!
    static var jobStatuses: [JobStatus]!
    static var applicationStatuses: [ApplicationStatus]!
    static var roles: [Role]!
    static var initialTokens: InitialTokens!
    
    static var userRole: NSNumber!
    static var jobSeeker: JobSeeker!
    static var profile: Profile!
    
    static var businesses = [Business]()
    static var workplaces = [Location]()
    static var jobs = [Job]()
    static var businessUsers = [BusinessUser]()
    static var jobSeekers = [JobSeeker]()
    static var applications = [Application]()
    static var newMessageCount = 0
    
    static var appsRefreshTime = DEFAULT_REFRESH_TIME
    static var appsUpdateCallback: (() -> Void)?
    
    static func clearData() {
        user = nil
        userRole = nil
        jobSeeker = nil
        profile = nil
        
        businesses.removeAll()
        workplaces.removeAll()
        jobs.removeAll()
        businessUsers.removeAll()
        jobSeekers.removeAll()
        applications.removeAll()
        
        stopTimer()
    }
    
    static func loadData(complete: ((Any?) -> Void)!) {
        
        var failed = false
        
        let loadFailure = { (error: Any?) in
            if !failed {
                failed = true
                complete(error)
            }
        }
        
        let loadSuccess = {
            if ( hours != nil &&
                contracts != nil &&
                sexes != nil &&
                nationalities != nil &&
                sectors != nil &&
                jobStatuses != nil &&
                applicationStatuses != nil &&
                roles != nil &&
                initialTokens != nil) {
                
                if (user.isJobSeeker()) {
                    
                    userRole = Role.ROLE_JOB_SEEKER_ID
                    
                    getJobSeeker() { error in
                        if error != nil {
                            complete(error)
                            return
                        }
                        getProfile() { error in
                            if error != nil {
                                complete(error)
                                return
                            }
                            getApplications() { error in
                                complete(error)
                                if error == nil {
                                    startTimer()
                                }
                                
                            }
                        }
                    }
                    
                } else if (user.isRecruiter()) {
                    
                    userRole = Role.ROLE_RECRUITER_ID
                    
                    getBusinesses() { error in
                        getApplications() { error in
                            complete(error)
                            if error == nil {
                                startTimer()
                            }
                        }
                    }
                } else {
                    complete(nil)
                }
            }
        }
        
        if initialTokens == nil {
            API.shared().loadHours() { (result, error) in
                if result != nil {
                    hours = result as! [Hours]
                    loadSuccess()
                } else {
                    loadFailure(error)
                }
            }
            
            API.shared().loadContracts() { (result, error) in
                if result != nil {
                    contracts = result as! [Contract]
                    loadSuccess()
                } else {
                    loadFailure(error)
                }
            }
            
            API.shared().loadSexes() { (result, error) in
                if result != nil {
                    sexes = result as! [Sex]
                    loadSuccess()
                } else {
                    loadFailure(error)
                }
            }
            
            API.shared().loadNationalities() { (result, error) in
                if result != nil {
                    nationalities = result as! [Nationality]
                    loadSuccess()
                } else {
                    loadFailure(error)
                }
            }
            
            API.shared().loadSectors() { (result, error) in
                if result != nil {
                    sectors = result as! [Sector]
                    loadSuccess()
                } else {
                    loadFailure(error)
                }
            }
            
            API.shared().loadJobStatuses() { (result, error) in
                if result != nil {
                    jobStatuses = result as! [JobStatus]
                    JobStatus.JOB_STATUS_OPEN_ID = getIdByName(jobStatuses, name: JobStatus.JOB_STATUS_OPEN)
                    JobStatus.JOB_STATUS_CLOSED_ID = getIdByName(jobStatuses, name: JobStatus.JOB_STATUS_CLOSED)
                    loadSuccess()
                } else {
                    loadFailure(error)
                }
            }
            
            API.shared().loadApplicationStatuses() { (result, error) in
                if result != nil {
                    applicationStatuses = result as! [ApplicationStatus]
                    ApplicationStatus.APPLICATION_CREATED_ID = getIdByName(applicationStatuses, name: ApplicationStatus.APPLICATION_CREATED)
                    ApplicationStatus.APPLICATION_ESTABLISHED_ID = getIdByName(applicationStatuses, name: ApplicationStatus.APPLICATION_ESTABLISHED)
                    ApplicationStatus.APPLICATION_DELETED_ID = getIdByName(applicationStatuses, name: ApplicationStatus.APPLICATION_DELETED)
                    loadSuccess()
                } else {
                    loadFailure(error)
                }
            }
            
            API.shared().loadRoles() { (result, error) in
                if result != nil {
                    roles = result as! [Role]
                    Role.ROLE_JOB_SEEKER_ID = getIdByName(roles, name: Role.ROLE_JOB_SEEKER)
                    Role.ROLE_RECRUITER_ID = getIdByName(roles, name: Role.ROLE_RECRUITER)
                    loadSuccess()
                } else {
                    loadFailure(error)
                }
            }
            
            API.shared().loadInitialTokens() { (result, error) in
                if result != nil {
                    initialTokens = result as! InitialTokens
                    loadSuccess()
                } else {
                    loadFailure(error)
                }
            }
        } else {
            loadSuccess()
        }
    }
    
    static func getIdByName(_ objects: [MJPObjectWithName], name: String!) -> NSNumber! {
        let filters = objects.filter { $0.name == name }
        return filters.count > 0 ? filters[0].id : nil
    }
    
    static func getNameByID(_ objects: [MJPObjectWithName], id: NSNumber!) -> String! {
        let filters = objects.filter { $0.id == id }
        return filters.count > 0 ? filters[0].name : nil
    }
    
    //================ jobseeker =============
    
    static func getJobSeeker(complete: ((Any?) -> Void)?) {
        if (user.isJobSeeker()) {
            API.shared().loadJobSeekerWithId(AppData.user.jobSeeker) { (result, error) in
                if result != nil {
                    jobSeeker = result as! JobSeeker
                }
                complete?(error)
            }
        } else {
            complete?(nil)
        }
    }
    
    static func getProfile(complete: ((Any?) -> Void)?) {
        if (jobSeeker?.profile != nil) {
            API.shared().loadJobProfileWithId(jobSeeker.profile) { (result, error) in
                if result != nil {
                    profile = result as! Profile
                }
                complete?(error)
            }
        } else {
            complete?(nil)
        }
    }

    //================ businesses =============
    
    static func getBusinesses(complete: ((Any?) -> Void)?) {
        API.shared().loadBusinesses() { (result, error) in
            if result != nil {
                businesses = result as! [Business]
            }
            complete?(error)
        }
    }
    
    static func updateBusiness(_ object: NSObject, complete: ((Business?, Any?) -> Void)?) {
        
        let _updateBusiness = { (newBusiness: Business) in
            var isNew = true
            
            for (index, business) in businesses.enumerated() {
                if business.id == newBusiness.id {
                    businesses[index] = newBusiness
                    isNew = false
                    break
                }
            }
            
            if isNew {
                businesses.insert(newBusiness, at: 0)
            }
            
            complete?(newBusiness, nil)
        }
        
        if let newBusiness = object as? Business {
            _updateBusiness(newBusiness)
        } else {
            API.shared().loadBusiness(object as! NSNumber) { (result, error) in
                if result != nil {
                    _updateBusiness(result as! Business)
                } else {
                    complete?(nil, error)
                }
            }
        }
    }
    
    static func removeBusiness(_ business: Business, complete: ((Any?) -> Void)?) {
        API.shared().deleteBusiness(business.id) { error in
            if error == nil {
                businesses = businesses.filter { $0.id != business.id }
            }
            complete?(error)
        }
    }
    
    //================ workplaces =============
    
    static func getWorkplaces(businessId: NSNumber, complete: ((Any?) -> Void)?) {
        API.shared().loadLocationsForBusiness(businessId) { (result, error) in
            if (result != nil) {
                workplaces = result as! [Location]
            }
            complete?(error)
        }
    }
    
    static func updateWorkplace(_ object: NSObject, complete: ((Location?, Any?) -> Void)?) {
        
        let _updateWorkplace = { (newWorkplace: Location) in
            var isNew = true
            
            for (index, location) in workplaces.enumerated() {
                if location.id == newWorkplace.id {
                    workplaces[index] = newWorkplace
                    isNew = false
                    break
                }
            }
            
            if isNew {
                workplaces.insert(newWorkplace, at: 0)
            }
            
            updateBusiness(newWorkplace.businessData, complete: nil)
            
            complete?(newWorkplace, nil)
        }
        
        if let newWorkplace = object as? Location {
            _updateWorkplace(newWorkplace)
        } else {
            API.shared().loadLocation(object as! NSNumber) { (result, error) in
                if (error == nil) {
                    _updateWorkplace(result as! Location)
                } else {
                    complete?(nil, error)
                }
            }
        }
    }
    
    static func removeWorkplace(_ workplace: Location, complete: ((Any?) -> Void)?) {
        API.shared().deleteLocation(workplace.id) { error in
            if error == nil {
                workplaces = workplaces.filter { $0.id != workplace.id }
                let businesses1 = businesses.filter { $0.id == workplace.businessData.id }
                if businesses1.count > 0 {
                    businesses1[0].locations = (businesses1[0].locations as! [NSNumber]).filter { $0 != workplace.id } as NSArray!
                }
            }
            complete?(error)
        }
    }
    
    //================ jobs =============
    
    static func getJobs(locationId: NSNumber!, complete: ((Any?) -> Void)?) {
        API.shared().loadJobsForLocation(locationId) { (result, error) in
            if result != nil {
                jobs = result as! [Job]
            }
            complete?(error)
        }
    }
    
    static func searchJobs(complete: ((Any?) -> Void)?) {
        API.shared().searchJobs() { (result, error) in
            if result != nil {
                jobs = result as! [Job]
            }
            complete?(error)
        }
    }
    
    static func updateJob(_ object: NSObject, complete: ((Job?, Any?) -> Void)?) {
        
        let _updateJob = { (newJob: Job) in
            var isNew = true
            
            for (index, job) in jobs.enumerated() {
                if job.id == newJob.id {
                    jobs[index] = newJob
                    isNew = false
                    break
                }
            }
            
            if isNew {
                jobs.insert(newJob, at: 0)
            }
            
            updateWorkplace(newJob.locationData, complete: nil)

            complete?(newJob, nil)
        }
        
        if let newJob = object as? Job {
            _updateJob(newJob)
        } else {
            API.shared().loadJob(object as! NSNumber) { (result, error) in
                if result != nil {
                    _updateJob(result as! Job)
                } else {
                    complete?(nil, error)
                }
            }
        }
    }
    
    static func removeJob(_ job: Job, complete: ((Any?) -> Void)?) {
        API.shared().deleteJob(job.id) { error in
            if error == nil {
                jobs = jobs.filter { $0.id != job.id }
                let workplaces1 = workplaces.filter { $0.id == job.locationData.id }
                if workplaces1.count > 0 {
                    workplaces1[0].jobs = (workplaces1[0].jobs as! [NSNumber]).filter { $0 != job.id } as NSArray!
                }
            }
            complete?(error)
        }
    }
    
    //================ jobseekers =============
    
    static func searchJobseekers(jobId: NSNumber, complete: ((Any?) -> Void)?) {
        API.shared().searchJobSeekersForJob(jobId) { (result, error) in
            if result != nil {
                jobSeekers = result as! [JobSeeker]
            }
            complete?(error)
        }
    }
    
    static func removeJobseeker(_ jobseekerId: NSNumber) {
        for (index, jobseeker) in jobSeekers.enumerated() {
            if jobseeker.id == jobseekerId {
                jobSeekers.remove(at: index)
                break
            }
        }
    }
    
    //================ business users =============
    
    static func getBusinessUsers(businessId: NSNumber, complete: ((Any?) -> Void)?) {
        API.shared().loadBusinessUsers(businessId) { (result, error) in
            if result != nil {
                businessUsers = result as! [BusinessUser]
            }
            complete?(error)
        }
    }
    
    static func getBusinessUser(businessId: NSNumber, userId: NSNumber, complete: ((BusinessUser?, Any?) -> Void)?) {
        API.shared().getBusinessUser(businessId: businessId, businessUserId: userId) { (result, error) in
            let newBusinessUser = result as? BusinessUser
            if newBusinessUser != nil {
                var isNew = true
                for (index, user) in businessUsers.enumerated() {
                    if user.id == newBusinessUser?.id {
                        businessUsers[index] = newBusinessUser!
                        isNew = false
                        break
                    }
                }
                
                if isNew {
                    businessUsers.append(newBusinessUser!)
                }
            }
            
            complete?(newBusinessUser, error)
        }
    }
    
    static func removeBusinessUser(_ userId: NSNumber) {
        for (index, user) in businessUsers.enumerated() {
            if user.id == userId {
                businessUsers.remove(at: index)
                break
            }
        }
    }
    
    //================ applications =============
    
    private static var timer: Timer?
    private static var time = 0
    
    private static func startTimer() {
        if timer == nil {
            time = 0
            timer = Timer.scheduledTimer(timeInterval: 1, target: self, selector: #selector(refreshApplications), userInfo: nil, repeats: true)
        }
    }
    
    private static func stopTimer() {
        if timer != nil {
            timer?.invalidate()
            timer = nil
        }
    }
    
    private static func getNewMessageCount() {
        newMessageCount = 0
        
        for application in applications {
            newMessageCount += application.getNewMessageCount()
        }
    }
    
    static func refreshApplications() {
        time += 1
        if time >= appsRefreshTime {
            time = 0
            getApplications() { error in
                if error == nil {
                    appsUpdateCallback?()
                }
            }
        }
    }
    
    static func getApplications(complete: ((Any?) -> Void)?) {
        API.shared().loadApplications() { (result, error) in
            if result != nil {
                applications = result as! [Application]
                getNewMessageCount()
            }
            complete?(error)
        }
    }
    
    static func getApplication(_ id: NSNumber, complete: ((Application?, Any?) -> Void)?) {
        API.shared().loadApplicationWithId(id) { (result, error) in
            let newApplication = result as? Application
            if newApplication != nil {
                var isNew = true
                for (index, application) in applications.enumerated() {
                    
                    if application.id == newApplication?.id {
                        applications[index] = newApplication!
                        isNew = false
                        break
                    }
                }
                
                if isNew {
                    applications.insert(newApplication!, at: 0)
                }
                
                getNewMessageCount()
            }

            complete?(newApplication, error)
        }
    }
    
}

File: android/app/src/main/java/com/myjobpitch/utils/AppData.java
package com.myjobpitch.utils;

import android.support.v7.app.AppCompatActivity;

import com.myjobpitch.pages.MainActivity;
import com.myjobpitch.api.MJPAPIObject;
import com.myjobpitch.api.MJPApi;
import com.myjobpitch.api.MJPObjectWithName;
import com.myjobpitch.api.auth.User;
import com.myjobpitch.api.data.Application;
import com.myjobpitch.api.data.ApplicationStatus;
import com.myjobpitch.api.data.Business;
import com.myjobpitch.api.data.BusinessUser;
import com.myjobpitch.api.data.Contract;
import com.myjobpitch.api.data.Hours;
import com.myjobpitch.api.data.InitialTokens;
import com.myjobpitch.api.data.Job;
import com.myjobpitch.api.data.JobProfile;
import com.myjobpitch.api.data.Jobseeker;
import com.myjobpitch.api.data.JobStatus;
import com.myjobpitch.api.data.Workplace;
import com.myjobpitch.api.data.Message;
import com.myjobpitch.api.data.Nationality;
import com.myjobpitch.api.data.Role;
import com.myjobpitch.api.data.Sector;
import com.myjobpitch.api.data.Sex;

import java.util.ArrayList;
import java.util.List;
import java.util.Timer;
import java.util.TimerTask;

public class AppData {

    public static final int API_VERSION = 7;
    public static final boolean PRODUCTION = false;

    public static final int DEFAULT_REFRESH_TIME = 30;
    public static final int MESSAGE_REFRESH_TIME = 5;

    public static boolean existProfile = false;

    public static final int JOBSEEKER = 1;
    public static final int RECRUITER = 2;


    public static final int REQUEST_IMAGE_CAPTURE = 10000;
    public static final int REQUEST_IMAGE_PICK = 10001;
    public static final int REQUEST_GOOGLE_DRIVE = 10002;
    public static final int REQUEST_DROPBOX = 10003;


    public static final String KEY_EMAIL = "email";
    public static final String KEY_SERVER_URL = "server_url";
    public static final String KEY_TOKEN = "token";


    public static User user;

    public static List<Hours> hours;
    public static List<Contract> contracts;
    public static List<Sex> sexes;
    public static List<Nationality> nationalities;
    public static List<Sector> sectors;
    public static List<JobStatus> jobStatuses;
    public static List<ApplicationStatus> applicationStatuses;
    public static List<Role> roles;
    public static InitialTokens initialTokens;

    public static int userRole = -1;
    public static Jobseeker jobseeker;
    public static JobProfile profile;

    public static List<Business> businesses = new ArrayList<>();
    public static List<Workplace> workplaces = new ArrayList<>();
    public static List<Job> jobs = new ArrayList<>();
    public static List<BusinessUser> businessUsers = new ArrayList<>();
    public static List<Jobseeker> jobseekers = new ArrayList<>();
    public static List<Application> applications = new ArrayList<>();
    public static int newMessageCount = 0;

    public static int appsRefreshTime = DEFAULT_REFRESH_TIME;
    //    public static var appsUpdateCallback: (() -> Void)?


    public static void clearData() {
        MJPApi.shared().clearToken();

        existProfile = false;

        user = null;
        userRole = -1;
        jobseeker = null;
        profile = null;

        businesses = new ArrayList<>();
        workplaces = new ArrayList<>();
        jobs = new ArrayList<>();
        businessUsers = new ArrayList<>();
        jobseekers = new ArrayList<>();
        applications = new ArrayList<>();

        stopTimer();
    }

    public static void loadData() {
        user = MJPApi.shared().getUser();
        if (user.isJobseeker()) {
            Jobseeker jobseeker = MJPApi.shared().get(Jobseeker.class, user.getJob_seeker());
            existProfile = jobseeker.getProfile() != null;
        }

        if (initialTokens == null) {
            hours = MJPApi.shared().get(Hours.class);
            contracts = MJPApi.shared().get(Contract.class);
            sexes = MJPApi.shared().get(Sex.class);
            nationalities = MJPApi.shared().get(Nationality.class);
            sectors =  MJPApi.shared().get(Sector.class);
            jobStatuses = MJPApi.shared().get(JobStatus.class);
            applicationStatuses = MJPApi.shared().get(ApplicationStatus.class);
            roles = MJPApi.shared().get(Role.class);
            initialTokens = MJPApi.shared().getInitialTokens();

            JobStatus.OPEN_ID = getIdByName(jobStatuses, JobStatus.OPEN);
            JobStatus.CLOSED_ID = getIdByName(jobStatuses, JobStatus.CLOSED);
            ApplicationStatus.CREATED_ID = getIdByName(applicationStatuses, ApplicationStatus.CREATED);
            ApplicationStatus.ESTABLISHED_ID = getIdByName(applicationStatuses, ApplicationStatus.ESTABLISHED);
            ApplicationStatus.DELETED_ID = getIdByName(applicationStatuses, ApplicationStatus.DELETED);
            Role.JOB_SEEKER_ID = getIdByName(roles, Role.JOB_SEEKER);
            Role.RECRUITER_ID = getIdByName(roles, Role.RECRUITER);
        }

        if (user.isJobseeker()) {
            userRole = Role.JOB_SEEKER_ID;
            getJobseeker();
            getProfile();
            if (profile != null) {
                getApplications();
                startTimer();
            }
        } else if (user.isRecruiter()) {
            userRole = Role.RECRUITER_ID;
            getBusinesses();
            getApplications();
            startTimer();
        }
    }


    public static <T extends MJPObjectWithName> int getIdByName(List<T> objects, String name) {
        for (T obj : objects) {
            if (obj.getName().equals(name)) {
                return obj.getId();
            }
        }
        return -1;
    }

    public static <T extends MJPObjectWithName> String getNameById(List<T> objects, Integer id) {
        T obj = getObjById(objects, id);
        return obj != null ? obj.getName() : null;
    }

    public static <T extends MJPAPIObject> T getObjById(List<T> objects, Integer id) {
        for (T obj : objects) {
            if (obj.getId().equals(id)) {
                return obj;
            }
        }
        return null;
    }

    public static <T extends MJPAPIObject> void updateObj(List<T> objects, T updatedObj) {
        T obj = getObjById(objects, updatedObj.getId());
        if (obj != null) {
            objects.set(objects.indexOf(obj), updatedObj);
        } else {
            objects.add(0, updatedObj);
        }
    }

    //================ menu =============

    static void getJobseeker() {
        if (user.isJobseeker()) {
            jobseeker = MJPApi.shared().get(Jobseeker.class, user.getJob_seeker());
        }
    }

    static void getProfile() {
        if (jobseeker.getProfile() != null) {
            profile = MJPApi.shared().get(JobProfile.class, jobseeker.getProfile());
        }
    }

    //================ businesses =============

    public static void getBusinesses() {
        businesses = MJPApi.shared().getUserBusinesses();
    }

    public static void updateBusiness(Business busienss) {
        updateObj(businesses, busienss);
    }

    public static void getBusiness(Integer businessId) {
        Business business = MJPApi.shared().getUserBusiness(businessId);
        updateBusiness(business);
    }

    public void removeBusiness(Integer businessId) {
        Business obj = getObjById(businesses, businessId);
        if (obj != null) {
            businesses.remove(obj);
        }
    }

    //================ workplaces =============

    public static void getWorkplaces(Integer businessId) {
        workplaces = MJPApi.shared().getUserWorkplaces(businessId);
    }

    public static void updateWorkplace(Workplace workplace) {
        updateObj(workplaces, workplace);
        updateBusiness(workplace.getBusiness_data());
    }

    public static void getWorkplace(Integer workplaceId) {
        Workplace workplace = MJPApi.shared().getUserWorkplace(workplaceId);
        updateWorkplace(workplace);
    }

    public void removeWorkplace(Integer workplaceId) {
        Workplace obj = getObjById(workplaces, workplaceId);
        if (obj != null) {
            workplaces.remove(obj);
        }
    }

    //================ jobs =============

    public static void getJobs(Integer workplaceId) {
        jobs = MJPApi.shared().getUserJobs(workplaceId);
    }

    public static void searchJobs() {
        jobs = MJPApi.shared().get(Job.class);
    }

    public static void updateJob(Job job) {
        updateObj(jobs, job);
        updateWorkplace(job.getLocation_data());
    }

    public static void getJob(Integer jobId) {
        Job job = MJPApi.shared().getUserJob(jobId);
        updateJob(job);
    }

    public void removeJob(Integer jobId) {
        Job obj = getObjById(jobs, jobId);
        if (obj != null) {
            jobs.remove(obj);
        }
    }

    //================ jobseekers =============

    public static void searchJobseekers(Integer jobId) {
        jobseekers = MJPApi.shared().get(Jobseeker.class, "job=" + jobId);
    }

    public void removeJobseeker(Integer jobseekerId) {
        Jobseeker obj = getObjById(jobseekers, jobseekerId);
        if (obj != null) {
            jobseekers.remove(obj);
        }
    }

    //================ business users =============

    public static void getBusinessUsers(Integer businessId) {
        businessUsers = MJPApi.shared().getBusinessUsers(businessId);
    }

    public static void getBusinessUser(Integer businessId, Integer userId) {
        BusinessUser obj = MJPApi.shared().getBusinessUser(businessId, userId);
        updateObj(businessUsers, obj);
    }

    public void removeBusinessBuser(Integer userId) {
        BusinessUser obj = getObjById(businessUsers, userId);
        if (obj != null) {
            businessUsers.remove(obj);
        }
    }

    //================ applications =============

    static Timer timer;
    static int time = 0;

    public static void startTimer() {
        if (timer == null) {
            time = 0;
            timer = new Timer();
            timer.schedule(new TimerTask() {
                @Override
                public void run() {
                    time++;
                    if (time >= appsRefreshTime) {
                        time = 0;
                        getApplications();
                    }
                }
            }, 0, 1000);
        }
    }

    static void stopTimer() {
        if (timer != null) {
            timer.cancel();
            timer = null;
        }
    }

    static void getNewMessageCount() {
        newMessageCount = 0;

        for (Application app : applications) {
            int count = 0;
            for (int i = app.getMessages().size() - 1; i >= 0; i--) {
                Message message = app.getMessages().get(i);
                if (message.getRead()) {
                    break;
                }
                if (message.getFrom_role() == userRole) {
                    break;
                }
                count++;
            }

            newMessageCount += count;
        }
    }

    public static void getApplications() {
        applications = MJPApi.shared().get(Application.class);
        getNewMessageCount();
    }

    public static void getApplication(Integer id) {
        Application obj = MJPApi.shared().get(Application.class, id);
        updateObj(applications, obj);
        getNewMessageCount();
    }

}
